File: frontend/src/components/provider/providerDisplayName.test.js
/**
 * OGC-1223 FR-11: one helper renders a provider's name wherever a screen has
 * room for a single string, dropping a missing part cleanly rather than
 * leaving a leading space or a stray comma.
 */

import {
  providerDisplayName,
  titledProviderName,
  titledProviderNameFamilyFirst,
} from "./providerDisplayName";

describe("titledProviderName", () => {
  it("reads title, given name, family name", () => {
    expect(titledProviderName("Dr", "John", "Kila")).toBe("Dr John Kila");
  });

  it("leaves no leading space when there is no title", () => {
    expect(titledProviderName(null, "John", "Kila")).toBe("John Kila");
    expect(titledProviderName("", "John", "Kila")).toBe("John Kila");
    expect(titledProviderName("  ", "John", "Kila")).toBe("John Kila");
  });

  it("leaves no double space when a name part is missing", () => {
    expect(titledProviderName("Dr", null, "Kila")).toBe("Dr Kila");
    expect(titledProviderName("Dr", "John", null)).toBe("Dr John");
  });

  it("is empty rather than blank when nothing is known", () => {
    expect(titledProviderName(null, null, null)).toBe("");
  });

  it("trims what it is given", () => {
    expect(titledProviderName(" Dr ", " John ", " Kila ")).toBe("Dr John Kila");
  });
});

describe("titledProviderNameFamilyFirst", () => {
  it("keeps the title with the given name", () => {
    expect(titledProviderNameFamilyFirst("Dr", "John", "Kila")).toBe(
      "Kila, Dr John",
    );
  });

  it("drops the comma when there is no given name", () => {
    expect(titledProviderNameFamilyFirst(null, null, "Kila")).toBe("Kila");
  });

  it("reads as before when there is no title", () => {
    expect(titledProviderNameFamilyFirst(null, "John", "Kila")).toBe(
      "Kila, John",
    );
  });
});

describe("providerDisplayName", () => {
  it("prefers the resolved abbreviation over the raw code", () => {
    expect(
      providerDisplayName({
        titleCode: "DR",
        titleAbbreviation: "Dr",
        firstName: "John",
        lastName: "Kila",
      }),
    ).toBe("Dr John Kila");
  });

  it("falls back to the code when the title is no longer configured", () => {
    expect(
      providerDisplayName({
        titleCode: "HEO",
        firstName: "Anna",
        lastName: "Wanpis",
      }),
    ).toBe("HEO Anna Wanpis");
  });

  it("is empty for no provider at all", () => {
    expect(providerDisplayName(null)).toBe("");
  });
});

File: src/test/java/org/openelisglobal/provider/service/ProviderDisplayNameTest.java
package org.openelisglobal.provider.service;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

/**
 * OGC-1223 FR-11: one helper renders a provider's name wherever a screen has
 * room for a single string, and a missing part never leaves a stray space or
 * separator behind.
 */
public class ProviderDisplayNameTest {

    @Test
    public void titledName_readsTitleFirstNameLastName() {
        assertEquals("Dr John Kila", ProviderDisplayName.titled("Dr", "John", "Kila"));
    }

    @Test
    public void noTitle_leavesNoLeadingSpace() {
        assertEquals("John Kila", ProviderDisplayName.titled(null, "John", "Kila"));
        assertEquals("John Kila", ProviderDisplayName.titled("", "John", "Kila"));
        assertEquals("John Kila", ProviderDisplayName.titled("   ", "John", "Kila"));
    }

    @Test
    public void noFirstName_leavesNoDoubleSpace() {
        assertEquals("Dr Kila", ProviderDisplayName.titled("Dr", null, "Kila"));
    }

    @Test
    public void onlyATitle_rendersJustTheTitle() {
        assertEquals("Dr", ProviderDisplayName.titled("Dr", null, null));
    }

    @Test
    public void nothingAtAll_isEmptyRatherThanSpaces() {
        assertEquals("", ProviderDisplayName.titled(null, null, null));
    }

    @Test
    public void surroundingWhitespaceIsTrimmedNotPreserved() {
        assertEquals("Dr John Kila", ProviderDisplayName.titled(" Dr ", " John ", " Kila "));
    }

    @Test
    public void familyFirst_putsTheTitleWithTheGivenName() {
        assertEquals("Kila, Dr John", ProviderDisplayName.titledFamilyFirst("Dr", "John", "Kila"));
    }

    @Test
    public void familyFirst_withoutATitle_readsAsBefore() {
        assertEquals("Kila, John", ProviderDisplayName.titledFamilyFirst(null, "John", "Kila"));
    }

    @Test
    public void familyFirst_withoutAGivenName_dropsTheComma() {
        assertEquals("Kila", ProviderDisplayName.titledFamilyFirst(null, null, "Kila"));
    }

    @Test
    public void familyFirst_withoutAFamilyName_stillReads() {
        assertEquals("Dr John", ProviderDisplayName.titledFamilyFirst("Dr", "John", null));
    }
}
